dstask: init at 0.18

stianlagstad opened the pull request:
###### Motivation for this change

I'd like to still use https://github.com/naggie/dstask after switching from Ubuntu to NixOS.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

wamserma commented on pkgs/applications/misc/dstask/default.nix:
```diff
@@ -0,0 +1,26 @@
+  version = "0.17";
+
+  src = fetchurl {
+    url = "https://github.com/naggie/dstask/releases/download/${version}/dstask-linux-amd64";
```
Please make the platform/arch a parameter of the url and/or restrict the platform accordingly with a proper attribute in the meta set.

stianlagstad replied:
Thank you for the feedback. Is this latest commit doing it correctly? https://github.com/NixOS/nixpkgs/pull/87383/commits/586ae5a10cd8b79f2c406a3e36c5ff2cacd91ffe

wamserma commented:
```
undefined variable 'stianlagstad' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/packet-spot-eval-1/pkgs/applications/misc/dstask/default.nix:24:39
```

Please also add yourself to https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix

symphorien commented:
Please consider building from source instead. Go packages should be rather easy to package, see https://nixos.org/nixpkgs/manual/#sec-language-go

stianlagstad commented:
@symphorien Thanks for responding. I'm giving it a shot. Here's my current attempt:
```
{ stdenv, buildGoModule, fetchFromGitHub }:

buildGoModule rec {
  pname = "dstask";
  version = "0.17";

  src = fetchFromGitHub {
    owner = "naggie";
    repo = "dstask";
    rev = "{$version}";
    sha256 = "1z00f7ar8a0ahspgc240iq7cka1anls383d1fv3i17sb6ajmp6rc"; # I got this from `nix-prefetch-url --unpack https://github.com/naggie/dstask/archive/47cb278daccc520d73a9838462d60f514aedd87f.tar.gz`
  };

  modSha256 = "1z00f7ar8a0ahspgc240iq7cka1anls383d1fv3i17sb6ajmp6rc"; # I have no idea where to get this from. For now just copied in the same as the above.

  subPackages = [ "." ];

  meta = with stdenv.lib; {
    description = "Command line todo list with super-reliable git sync";
    homepage = "https://github.com/naggie/dstask";
    license = licenses.mit;
    maintainers = with maintainers; [ stianlagstad ];
    platforms = platforms.linux;
  };
}
```

When running `nix-build -A dstask -K` (from the root directory) I get this:
```
> nix-build -A dstask -K
these derivations will be built:
  /nix/store/vpayn6fjjc03zyly1dkm1mai09vbzhs0-source.drv
  /nix/store/nxmv80pajayd7p8qdkah0zmqsz7dnzkq-dstask-0.17-go-modules.drv
  /nix/store/1mdr6071k78krbahy0hqd0lnzh1amnjh-dstask-0.17.drv
building '/nix/store/vpayn6fjjc03zyly1dkm1mai09vbzhs0-source.drv'...

trying https://github.com/naggie/dstask/archive/{$version}.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   125  100   125    0     0    291      0 --:--:-- --:--:-- --:--:--   291
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
curl: (22) The requested URL returned error: 404 Not Found
error: cannot download source from any mirror
note: keeping build directory '/tmp/nix-build-source.drv-6'
builder for '/nix/store/vpayn6fjjc03zyly1dkm1mai09vbzhs0-source.drv' failed with exit code 1
cannot build derivation '/nix/store/1mdr6071k78krbahy0hqd0lnzh1amnjh-dstask-0.17.drv': 1 dependencies couldn't be built
error: build of '/nix/store/1mdr6071k78krbahy0hqd0lnzh1amnjh-dstask-0.17.drv' failed
```

I don't understand why it's trying to fetch "https://github.com/naggie/dstask/archive/{$version}.tar.gz" instead of "https://github.com/naggie/dstask/archive/0.17.tar.gz". Could you give me a pointer?

symphorien commented:
I think you meant `"${version}";` instead of `"{$version}";`

About hashes, replace them with zeroes, and nix will give you the right hash in the error message.

stianlagstad commented:
Thanks! I now tried this:
```
{ stdenv, buildGoModule, fetchFromGitHub }:

buildGoModule rec {
  pname = "dstask";
  version = "0.17";

  src = fetchFromGitHub {
    owner = "naggie";
    repo = "dstask";
    rev = "${version}";
    sha256 = "1z00f7ar8a0ahspgc240iq7cka1anls383d1fv3i17sb6ajmp6rc";
  };

  modSha256 = "1ydzqg8p2d514sdb34b2p6k1474nr1drrn3gay2cpyhrj5l51cj3";

  subPackages = [ "." ];

  meta = with stdenv.lib; {
    description = "Command line todo list with super-reliable git sync";
    homepage = "https://github.com/naggie/dstask";
    license = licenses.mit;
    maintainers = with maintainers; [ stianlagstad ];
    platforms = platforms.linux;
  };
}
```
Here's what happen when I build and look in the result directory:
```
> nix-build -A dstask -K                                         
these derivations will be built:
  /nix/store/kjg11z4j8cspzg03jg52bdxk0nxhghr9-dstask-0.17.drv
building '/nix/store/kjg11z4j8cspzg03jg52bdxk0nxhghr9-dstask-0.17.drv'...
unpacking sources
unpacking source archive /nix/store/9kxb3dig60vxwf52gdz498kbh7b5w6n8-source
source root is source
patching sources
configuring
building
Building subPackage ./.
go: downloading golang.org/x/sys v0.0.0-20181213200352-4d1cda033e06
go: downloading gopkg.in/yaml.v2 v2.2.2
go: downloading github.com/gofrs/uuid v3.2.0+incompatible
runtime/cgo
net
github.com/gofrs/uuid
golang.org/x/sys/unix
gopkg.in/yaml.v2
os/user
github.com/naggie/dstask
installing
post-installation fixup
find: '/nix/store/599xncfk2ayd40v2v3ns851nv1jrv27c-dstask-0.17/bin': No such file or directory
shrinking RPATHs of ELF executables and libraries in /nix/store/599xncfk2ayd40v2v3ns851nv1jrv27c-dstask-0.17
strip is /nix/store/3b3ighb83nhifa1v4n7855hlbdl1mhf9-binutils-2.31.1/bin/strip
patching script interpreter paths in /nix/store/599xncfk2ayd40v2v3ns851nv1jrv27c-dstask-0.17
checking for references to /build/ in /nix/store/599xncfk2ayd40v2v3ns851nv1jrv27c-dstask-0.17...
/nix/store/599xncfk2ayd40v2v3ns851nv1jrv27c-dstask-0.17
> ls -ahl /nix/store/599xncfk2ayd40v2v3ns851nv1jrv27c-dstask-0.17
total 980K
dr-xr-xr-x    2 root root   4.0K Jan  1  1970 .
drwxrwxr-t 2190 root nixbld 972K May  9 19:55 ..
```
Note the `find: '/nix/store/599xncfk2ayd40v2v3ns851nv1jrv27c-dstask-0.17/bin': No such file or directory` there. Does that mean that the build of https://github.com/naggie/dstask/ outputs the binary in another folder or something?

symphorien commented:
Honestly, I'm not very familiar with go. Maybe we can wait for someone more knowledgeable or you can ask on irc or on discourse ?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/help-building-a-go-package/7117/1

stianlagstad commented:
Thanks again @symphorien . I've asked for help [on the nixos discourse](https://discourse.nixos.org/t/help-building-a-go-package/7117).

stianlagstad commented:
Thanks to foxit64 [here](https://github.com/naggie/dstask/issues/19) the build now works!

stianlagstad commented:
[The build error](https://gist.github.com/GrahamcOfBorg/c7b0ee7f3ad134799742c15990e818a2) says:

> modSha256 is deprecated and will be removed in the next release (20.09), use vendorSha256 instead

I tried building locally just using `vendorSha256` instead of `modSha256` but that gave the error:

> error: anonymous function at /nix/var/nix/profiles/per-user/root/channels/nixos/pkgs/development/go-modules/generic/default.nix:3:1 called without required argument 'modSha256', at /home/stian/devp/nixpkgs/pkgs/applications/misc/dstask/default.nix:3:1

Does that mean that `buildGoModule` itself need to be changed for the build to pass?

foxit64 commented:
@stianlagstad 
Try it. I'm seeing this build error for the first time.

stianlagstad commented:
Looking at [go-modules/generic/default.nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/go-modules/generic/default.nix) I saw [this recent change](https://github.com/NixOS/nixpkgs/commit/9761128d2da7bf4d878982918242e43ae28f9b94). I've [updated the PR](https://github.com/NixOS/nixpkgs/pull/87383/commits/311de3ecdc65dec83f16299d0d93551e744b2c3e) to use `vendorSha256` instead of `modSha256`.

stianlagstad commented:
I see that [these](https://github.com/NixOS/nixpkgs/pull/87383/checks?check_run_id=702688100) [two](https://github.com/NixOS/nixpkgs/pull/87383/checks?check_run_id=702688091) checks fail with:

> vendor folder exists, please set 'vendorSha256=null;' or 'deleteVendor=true;' in your expression

That's from [this part of the nix expression](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/go-modules/generic/default.nix#L87):
```
    buildPhase = args.modBuildPhase or ''
      runHook preBuild
      if [ ${deleteFlag} == "true" ]; then
        rm -rf vendor
      fi
      if [ -e vendor ]; then
        echo "vendor folder exists, please set 'vendorSha256=null;' or 'deleteVendor=true;' in your expression"
        exit 10
      fi
      go mod vendor
      mkdir -p vendor
      runHook postBuild
    '';
```

I don't understand enough of this yet to know what to do. @naggie are you able to assist?

naggie commented:
I think `vendorSha256=null` should do it, as I do vendor my go modules (which means storing the thirdparty code in the repository)

I vendor the modules to make sure it's possible to build dstask even if the thirdparty dependencies are removed/changed from github.

naggie commented:
Also I wonder if we need to add `-mod=vendor` to the `buildFlagsArray`

stianlagstad commented:
@naggie Thanks! Looks like I messed up that build flag [in my change](https://github.com/NixOS/nixpkgs/pull/87383/commits/85388e430f844620cd2f7e8c7de1651dcf9ed643) though. The build log of the new failure:
> flag provided but not defined: -mod
(from <https://logs.nix.ci/?key=nixos/nixpkgs.87383&attempt_id=617729ef-59b3-437b-af72-45c4176daa3f>)

naggie commented:
I think that needs its own new line, else it's probably parsed as an ldflag:


```
 buildFlagsArray = [ ''-ldflags=-w -s
    -mod=vendor
    -X "github.com/naggie/dstask.VERSION=${version}"
    -X "github.com/naggie/dstask.GIT_COMMIT=${rev}"
  '' ];
```

foxit64 reviewed (approved):
LGTM

stianlagstad commented:
Thanks again @naggie and @foxit64 .

@foxit64 : Do you know if there's anything else that needs to be done to move the PR forward? Or do I just need to be a little more patient? :)

06kellyjac commented on pkgs/applications/misc/dstask/default.nix:
```diff
@@ -0,0 +1,31 @@
+
+  meta = with stdenv.lib; {
+    description = "Command line todo list with super-reliable git sync";
+    homepage = "https://github.com/naggie/dstask";
```
FYI, you can put `src.meta.homepage` to get the same value :slightly_smiling_face: 

https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/audio/adlplug/default.nix#L38
https://nixos.org/nixos/packages.html?channel=nixos-20.03&query=adlplug

06kellyjac commented on pkgs/applications/misc/dstask/default.nix:
```diff
@@ -0,0 +1,31 @@
+
+  src = fetchFromGitHub {
+    owner = "naggie";
+    repo = "dstask";
```
```suggestion
    repo = pname;
```

06kellyjac commented on pkgs/applications/misc/dstask/default.nix:
```diff
@@ -0,0 +1,31 @@
+{ stdenv, buildGoModule, fetchFromGitHub }:
+
+buildGoModule rec {
+  pname = "dstask";
+  version = "v0.18";
+  rev = "d687a0f9aec634a808cfa0df4aecb06ab4f57732";
+
+  src = fetchFromGitHub {
+    owner = "naggie";
+    repo = "dstask";
+    rev = "${version}";
+    sha256 = "16z5zlfj955pzsj0l58835slvpchdaq2vbyx2fjzi6y9xn1z2nd1";
+  };
```
```suggestion
  pname = "dstask";
  version = "0.18";

  src = fetchFromGitHub {
    owner = "naggie";
    repo = pname;
    rev = "v${version}";
    sha256 = "16z5zlfj955pzsj0l58835slvpchdaq2vbyx2fjzi6y9xn1z2nd1";
  };
```

`repo` is the same as `pname`

pkg versions shouldn't include "v"
(`buildFlagsArray` also needs updating)

Does `rev` in `buildGoModule` do anything? I can't find it there & I've not seen it used before.
https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/go-modules/generic/default.nix

06kellyjac replied:
Edit: ah, `rev` is for the `GIT_COMMIT` flag

dR3b replied:
Yes and for the command output: `dstask version`

timokau replied:
Regarding the `repo = pname` thing: I actually think this is an antipattern. `repo` happens to be the same as `panme`, but the two things are not the same concept and don't necessarily change together.

Its pointless to have a nitpick war about this since we don't have an established standard. So no need to change anything. Just couldn't resist mentioning it ;)

06kellyjac commented on pkgs/applications/misc/dstask/default.nix:
```diff
@@ -0,0 +1,31 @@
+  vendorSha256 = null;
+
+  buildFlagsArray = [ ''-ldflags=-w -s
+    -X "github.com/naggie/dstask.VERSION=${version}"
```
I should have included this in the original suggestion. I made it needlessly difficult :sweat_smile: 
```suggestion
    -X "github.com/naggie/dstask.VERSION=v${version}"
```

stianlagstad replied:
Sorry about that, I missed the comment. Thank you very much for the help!

foxit64 reviewed (approved):
LGTM :+1:

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review-may-2019/3032/172

timokau commented on pkgs/applications/misc/dstask/default.nix:
```diff
@@ -0,0 +1,31 @@
+
+  buildFlagsArray = [ ''-ldflags=-w -s
+    -X "github.com/naggie/dstask.VERSION=v${version}"
+    -X "github.com/naggie/dstask.GIT_COMMIT=${rev}"
```
What are these flags doing? Why are they needed? What does `rev` specify and why is it different from the source `rev`?

stianlagstad replied:
@foxit64 Would you be able to assist me here?

naggie replied:
The ldflags reduce the executable size by stripping some debug stuff.

The variables are set so that the output of `dstask version` shows the git ref and the release version from github.

foxit64 replied:
thanks naggie for the explanation.

stianlagstad replied:
Thanks naggie!:)

timokau replied:
So basically `ref` is just the fully-qualified commit hash of the tag `version`? It seems likely that someone will forget to update that in the future, especially since it will not lead to runtime breakage. If it is really necessary to encode that information, the `ref` definition should be close to `version` (as it already is) and there should be a comment that explains how `ref` should be updated when updating the package.

I generally don't see much point in duplicating this information though, since the version should convey the same information right? If we manually have to maintain that, it makes updates a little bit more cumbersome. More importantly, all updates will need to be done manually since our automatic update infrastructure (https://github.com/ryantm/nixpkgs-update) won't know about this.

stianlagstad replied:
It would be very nice if the package could be automatically updated with [nixpkgs-update](https://github.com/ryantm/nixpkgs-update)! What do you think, @naggie? Would [this code](https://github.com/naggie/dstask/blob/2921be9d9c300d963b79f0d4ceb7795c396aa8d4/cmd/dstask.go#L333) crash if we didn't supply the `GIT_COMMIT` build flag?

timokau commented on pkgs/applications/misc/dstask/default.nix:
```diff
@@ -0,0 +1,31 @@
+    sha256 = "16z5zlfj955pzsj0l58835slvpchdaq2vbyx2fjzi6y9xn1z2nd1";
+  };
+
+  vendorSha256 = null;
```
Please add an inline comment explaining why this is needed. This may just be a link to the upstream author's explanation.

stianlagstad replied:
Thanks! I've added a comment explaining this.

Lassulus commented:
squashing/rebasing the commits into 2 would be nice:
- one for the maintainer addition
- one for the package

timokau reviewed (changes requested):
Sorry for holding this up even longer, but there are a few things I would like to see clarified. Its important to document everything as well as possible in the expression itself, since the next person looking at it will not have the context you have right now (even if the next person is you).

I think we're close to the finish line though. Once you have made all the changes, please squash your commits so that only two remain:

```
maintainers: add stianlagstad
dstask: init at 0.18
```
(in that order). Note that the format of the maintainers commit message needs to be amended and the version in the init commit needs to be changed to 0.18.

If you're not familiar with the workflow, have a look at https://about.gitlab.com/blog/2018/06/07/keeping-git-commit-history-clean/ (situation 3).

Looking forward to merging this :)

stianlagstad commented:
Thank you @Lassulus and @timokau ! I'll have to wait for help from either @foxit64 or @naggie to explain those build flags. I'll rewrite the commits as required after adding a comment explaining the build flags.

naggie commented:
> Sorry for holding this up even longer,

No problem, I'm glad the standards are so high

stianlagstad committed:
maintainers: add stianlagstad

stianlagstad committed:
dstask: init at 0.18

Thank you to: foxit64, naggie, timokau, Lassulus, 06kellyjac,
symphorien, wamserma.

stianlagstad commented:
I've updated the commits now. Thank you for the help so far, and please do feel free to comment further as I'm in no particular rush and would like to follow the standards.

timokau commented on pkgs/applications/misc/dstask/default.nix:
```diff
@@ -0,0 +1,41 @@
+  # Ref <https://github.com/NixOS/nixpkgs/pull/87383#discussion_r432097657>
+  buildFlagsArray = [ ''
+    -ldflags=-w -s
+    -X "github.com/naggie/dstask.VERSION=v${version}"
+    -X "github.com/naggie/dstask.GIT_COMMIT=${rev}"
```
How about this:

```suggestion
    -X "github.com/naggie/dstask.VERSION=${version}"
    -X "github.com/naggie/dstask.GIT_COMMIT=v${version}"
```

I don't see why there should be a `v` prefix for the version. At the same time, `v${version}` is the name of the tag and therefore a valid specifier of a git commit.

naggie replied:
Seems like a good change as it would make the output of dstask version consistent with the official binary, which currently outputs:

```
Version: v0.18
Git commit: d687a0f9aec634a808cfa0df4aecb06ab4f57732
Build date: Wed 13 May 18:34:38 BST 2020

```

stianlagstad replied:
Thank you both again! I've updated the PR.

timokau replied:
@naggie it sound like you may be reading the diff the wrong way around. Previously it would have output

```
Version: v0.18
Git commit: d687a0f9aec634a808cfa0df4aecb06ab4f57732
```

but now it should be

```
Version: 0.18
Git commit: v0.18
```

The reasoning behind this is that we would like to avoid having to manually set the fully-qualified git revision of the release in our expression. That would be cumbersome and our automated tooling wouldn't update that. At the same time, since the version is tagged `v0.18` is also a valid reference to the git revision.

We've dropped the `v` prefix from `Version` since `0.18` is the version number while `v0.18` is the name of the tag.

Is that fine by you?

timokau commented on pkgs/applications/misc/dstask/default.nix:
```diff
@@ -0,0 +1,41 @@
+buildGoModule rec {
+  pname = "dstask";
+  version = "0.18";
+  rev = "d687a0f9aec634a808cfa0df4aecb06ab4f57732";
```
```suggestion
```

This is no longer used.

timokau reviewed (changes requested):
Looks good to me now, except that one line and pending approval of @naggie regarding the version/rev setting.

timokau reviewed (approved):
Looks great to me now. I'll merge this if and when @naggie approves. If I forget about this, please ping me again some time next week.

timokau commented:
Thanks everyone for your patience and responsiveness. I think the end-result is a high-quality package. I'll probably check it out some time, I have even thought about working on a taskwarrior alternative myself already.

Welcome to the team @stianlagstad :tada:

naggie commented:
Great, thanks to everyone, really appreciated!

timokau commented:
Pretty much. The build server periodically grabs the latest nixpkgs revision and builds all packages. Once that is done and some checks pass, the `nixos-unstable` channel is updated to that new revision. It'll probably take something between a day and a week until the package shows up in the channel.

You don't have to wait for that though. Nix makes it possible to build packages or even your whole system from any nixpkgs checkout, it doesn't have to be a channel. The packages that have already been built will transparently be pulled in from the binary caches, everything else will be built from source. Imagine a blend of ArchLinux and Gentoo. The point of channels is mostly to signal everyone "the binary channel for this revision is populated and the tests have passed".

naggie commented:
Oh I see, I like the sound of that. I prefer this to the AUR due to the review process, and the fact there's the binary cache means things are fast where possible.

foxit64 commented:
Congratulations to @stianlagstad for your first successful merge!

stianlagstad commented:
Thank you everyone for the help!

In case anyone else stumbles upon this and wants to install dstask before its available in a channel: See [Adding Custom Packages](https://nixos.org/nixos/manual/index.html#sec-custom-packages) in the NixOS manual.

naggie commented:
Cool, it's now listed: https://nixos.org/nixos/packages.html?channel=nixpkgs-unstable&query=dstask
